contracts/src/finance/vesting_wallet.rs

```diff
@@ -0,0 +1,617 @@
+
+    fn end(&self) -> U256 {
+        // SAFETY: both `start` and `duration` are stored as u64, so they cannot
+        // exceed `U256::MAX`
+        self.start() + self.duration()
+    }
+
+    #[selector(name = "released")]
+    fn released_eth(&self) -> U256 {
+        self._released.get()
+    }
+
+    #[selector(name = "released")]
+    fn released_erc20(&self, token: Address) -> U256 {
+        self._erc20_released.get(token)
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_eth(&self) -> U256 {
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        self.vested_amount_eth(block::timestamp()) - self.released_eth()
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_erc20(&self, token: Address) -> Result<U256, Self::Error> {
+        let vested = self.vested_amount_erc20(token, block::timestamp())?;
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        Ok(vested - self.released_erc20(token))
+    }
+
+    #[selector(name = "release")]
+    fn release_eth(&mut self) -> Result<(), Self::Error> {
+        let amount = self.releasable_eth();
+        let released = self
+            .released_eth()
+            .checked_add(amount)
+            .expect("should not exceed `U256::MAX` for `_released`");
+        self._released.set(released);
+
+        evm::log(EtherReleased { amount });
+
+        transfer_eth(self.ownable.owner(), amount)
+            .map_err(|_| Error::ReleaseEtherFailed(ReleaseEtherFailed {}))
+    }
+
+    #[selector(name = "release")]
+    fn release_erc20(&mut self, token: Address) -> Result<(), Self::Error> {
+        let amount = self.releasable_erc20(token)?;
+        let owner = self.ownable.owner();
+
+        self.safe_erc20.safe_transfer(token, owner, amount).map_err(|_| {
+            Error::ReleaseTokenFailed(ReleaseTokenFailed { token })
+        })?;
+
+        // SAFETY: total supply of a [`crate::token::erc20::Erc20`] cannot
+        // exceed `U256::MAX`.
+        self._erc20_released.setter(token).add_assign_unchecked(amount);
+
+        evm::log(ERC20Released { token, amount });
+
+        Ok(())
+    }
+
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_eth(&self, timestamp: u64) -> U256 {
+        let total_allocation = contract::balance()
+            .checked_add(self.released_eth())
+            .expect("total allocation should not exceed `U256::MAX`");
+
+        self.vesting_schedule(total_allocation, U64::from(timestamp))
+    }
+
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_erc20(
+        &self,
+        token: Address,
+        timestamp: u64,
+    ) -> Result<U256, Self::Error> {
+        if !Address::has_code(&token) {
+            return Err(InvalidToken { token }.into());
+        }
+
+        let call = IErc20::balanceOfCall { account: contract::address() };
+        let balance = RawCall::new()
```

Comment:
I believe [`Call`](https://docs.rs/stylus-sdk/latest/stylus_sdk/call/struct.Call.html) would be more convenient to use here. But you would need to `sol_interface!` macro instead of `sol!`.

Reply:
That's a great question!
I used `RawCall` in `SafeErc20` because most of the functions that need to be called may return bool value or nothing -- `RawCall` allows more advanced configurations for other contracts calls (like [setting return data length](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/contracts/src/token/erc20/utils/safe_erc20.rs#L369)). For `RawCall` contract's interface needs to be specified in `alloy_sol_types::sol!` macro.
On the other hand `Call` is more convenient to be used by a specific interface (the interface needs to be specified in `stylus_sdk::prelude::solidity_interface!` macro) - it calls a contract and casts the output to the specified one as you can see in this example:
```
use stylus_sdk::call::{Call, Error};
use stylus_sdk::{prelude::*, evm, msg, alloy_primitives::Address};
extern crate alloc;

sol_interface! {
    interface IService {
        function makePayment(address user) external payable returns (string);
    }
}

pub fn do_call(account: IService, user: Address) -> Result<String, Error> {
    let config = Call::new()
        .gas(evm::gas_left() / 2)       // limit to half the gas left
        .value(msg::value());           // set the callvalue

    account.make_payment(config, user)  // note the snake case
}
```

To summarise, I wanted to have a consistent calling within a smart contract implementation.